### ISSUE TITLE
Migrations are not properly being applied due to invalid foreign key constraints on gravitylog.device

### ISSUE THREAD

**Issue by thorrak** [thorrak/fermentrack#450]
IntegrityError: The row in table 'gravity_gravitylog' with primary key '1' has an invalid foreign key: gravity_gr...

```
IntegrityError: The row in table 'gravity_gravitylog' with primary key '1' has an invalid foreign key: gravity_gravitylog.device_id contains a value '2' that does not have a corresponding value in gravity_gravitysensor__old.id.
(7 additional frame(s) were not displayed)
...
  File "django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "django/db/migrations/executor.py", line 247, in apply_migration
    migration_recorded = True
  File "django/db/backends/sqlite3/schema.py", line 35, in __exit__
    self.connection.check_constraints()
  File "django/db/backends/sqlite3/base.py", line 331, in check_constraints
    bad_value, referenced_table_name, referenced_column_name

IntegrityError: The row in table 'gravity_gravitylog' with primary key '1' has an invalid foreign key: gravity_gravitylog.device_id contains a value '2' that does not have a corresponding value in gravity_gravitysensor__old.id.
```

**Comment by thorrak** [comment]
This actually looks like it may be due to the user having a bad SD card. Sticking a pin in it for now.

**Comment by thorrak** [comment]
Although the first user to report this may have had a bad SD card, there does appear to be a bug here. 

This should be fixed by https://github.com/thorrak/fermentrack/pull/454/commits/7fae5541d473509d032af13ea344e574360b1850

**Comment by uSlackr** [comment]
I upgrade to the latest and I continue t get this message: 
"Fermentrack has upgraded to a newer copy of Django which requires an additional step to complete. Click here to trigger this step and restart Fermentrack."

Mu upgrade.log in the log directory has this same error.

**Comment by thorrak** [comment]
Can you send me a copy of your SQLite database file? I keep trying to debug this, but none of my installs have it, which is making it difficult.

**Comment by thorrak** [comment]
You are a saint amongst men. I’ll download as soon as I get back to my laptop.

**Comment by thorrak** [comment]
Excellent news! Your database crashes on my end too. Now I can actually get to work fixing this, rather than throwing stuff at the wall and hoping something sticks.

Thank you!

**Comment by thorrak** [comment]
OK, I've managed to code something up that fixes _this_ specific problem at least. It should be on dev now.

**Comment by uSlackr** [comment]
Upgrading one now. Thanks!

**Comment by uSlackr** [comment]
that brought my site back with no apparent problem  

[upgrade.log](https://github.com/thorrak/fermentrack/files/4500375/upgrade.log)

**Comment by thorrak** [comment]
Awesome! Thank you again for providing that database - there was exactly 0 chance I would have stumbled into that solution otherwise.

**Comment by uSlackr** [comment]
Glad to help.  And thanks for the software and for all the Tilt love recently.
My second pi came back too!

**Comment by uSlackr** [comment]
There may still be an issue. I started a new log and it doesn't appear to be logging.  No graph or data file yet.  I'm still watching.

**Comment by thorrak** [comment]
When you say it doesn't appear to be logging, what kind of log are you looking at? A standalone gravity log, a combined temperature sensor and gravity log, or a log that is on an external website (that the external push support is pushing to)?


I ask, because there may actually be two separate issues --

1. Someone - not sure if it was you or someone else - had the same issue as before, because there were _two_ tables that had malformed foreign key constraints due to the way that Django 1.11 and earlier set them up. Apparently, the fix that I pushed before only works if there's only _one_ table. If you're having an issue on one pi but not the other, this may be cause. What I _think_ is the fix for this has been pushed to dev and should be available now.
2. There is a separate issue with Tilt support under Django 3.0 where it seems like some calls to log data can fail because Django implemented its own use of asyncio. I'm testing a fix for this now, and am tracking that issue as #456 . If the test goes well, it will be on dev shortly.

**Comment by uSlackr** [comment]
I'm not getting a graph, and I'm not seeing the csv files in the data dir.
I'll update again

**Comment by uSlackr** [comment]
I do however see the entry in the Gravityog table.

**Comment by thorrak** [comment]
This should be fixed by the pull request I just merged into Dev (#459)

Let me know if you have any issues!

**Comment by uSlackr** [comment]
I have a data file and graph after the latest patch

**Comment by thorrak** [comment]
In addition to the issues you ran into @uSlackr there appears to have been a _separate_ scenario where migrations generating new tables would fail under Django 3.0 due to failing foreign key checks. This should now be fixed following commit bc305c302ae0642bcefda9ee4b17e6f8dc5da1d6 which adds a new management command - `migrate_no_fk` - which runs migrations while disabling foreign key checks.

For most users this should only ever have to be run once, so I've also triggered a call to this from the `fix_sqlite_for_django_2` management command which can itself be triggered from within the Fermentrack UI.

**Comment by uSlackr** [comment]
I‘m upgrading now.  I’ll run it once done.

Thanks for the note.

BTW, been listening to HomeBrewDYI recently. Saw yu did a session with Colson.  I hoope to listen to it today.  Thanks for the work on this!

\\Greg
